branches/to-cmake/debian/fte.1
.TH FTE 1 
.SH NAME
fte \- Text editor
.SH SYNOPSIS
.B fte
.I " [[options] files ...]"
.BR
.B xfte
.I "[[options] files ...]"
.BR
.B vfte 
.I "[[options] files ...]"
.BR
.B nfte 
.I "[[options] files ...]"
.BR
.\" .I "[[options] files ...]"
.SH "DESCRIPTION"
This manual page documents briefly the
.BR "fte" .

This manual page was written for the Debian GNU/Linux distribution
(but may be used by others), because the original program does not
have a manual page.
Instead, it has documentation in html format; see below.
.PP
.B fte
is a powerful text editor which closely resembles Borlands programming
editors for M$\-DOS, but is far more flexible and configurable.
.SH OPTIONS
.BR fte
uses a command line syntax, which is a little bit different from
most of Debian GNU/Linux apps.
.TP
.B \-h \-? \--help
Help on command line options.
.TP
.B \-! 
Ignore any external configuration file and use internal configuration.
.TP
.BI \-l <line>[,<column>]
Go to line (and column) in next file on command line.
.TP
.BI \-# <line>[<,column>]
Same as \fB-l\fP. Do not use this option.
.TP
.BI \-m [<mode>]
Set mode for next files to <mode>. If just plain \fB\-m\fP,
then use default mode.
.TP
.BI \-C [<.cnf>]
Use specified configuration file (compiled). Plain \fB\-C\fP does the same as
\fB\-!\fP.
.TP
.BI \-D [<.dsk>]
Load/save desktop from file <.dsk>. If no argument, disable
desktop load/save.
.TP
.BI \-H [<.his>]
Load/save history from file <.his>. If no argument, disable
history load/save.
.TP
.BI \-T <tagfile>
Load 'tagfile' at startup.
.TP
.BI \-t <tag>
Lookup tag 'tag' and display file containing it.
.TP
.B \-\-
The rest of the arguments are not options, but filenames.
.TP
.B \-+
The next argument is not an option even if starting with a '-'.
Useful when loading files that begin with a -.
.TP
.B \-noxmb
Do not use XMB functions in \fBxfte\fP - useful if you have incorrect locales
and you want to use just english fonts.
.TP
.B \-font
Select XWindow font to use with \fBxfte\fP. See \fBVIOFONT\fP also.

.SH ENVIRONMENTAL VARIABLES
This is list of variable which might change behavior of some \
internals of \fBfte\fP.
.TP
\fBVIOFONT\fP \- select font used by xfte 
.TP
\fBISOCONSOLE\fP \- when set, vfte is usign ascii symbols to draw tables. \
This is useful when iso latin charset is used on console.

.SH EXAMPLES
.TP
.BI "fte \-m" "BIN /usr/bin/vfte"
load /usr/bin/vfte in BIN mode
.TP
\fBfte \-m\fP\fIBIN\fP \fB\-+\fP \fI\-bla\-\fP
load file -bla- in BIN mode
.TP
.BI "fte \-l" "100,30 win.c"
go to (100,30) in win.c
.TP
.BI fte " window.cpp"
load file window.cpp
.TP
.BI "fte \-\-" " \-1 \-2 \-3 \-4 \-5 \-6"
load files \-1, \-2, \-3, \-4, \-5, \-6
.TP
.BI "fte \-D \-H" " fte.dsk fte.his"
disable desktop and history loading and saving and load files
fte.dsk and fte.his
.SH FILES
.I ~/.fterc
.br
.BI /usr/local/etc/ fte/system.fterc
.br
.BI /etc/ fte/system.fterc

Default configuration files. Includes all keybindings and menus
compiled with
.BR "cfte"(1)
from config files in \fB/usr/share/doc/\fP\fIfte/config-example.tar.gz\fP.
.SH "SEE ALSO"
.BR cfte (1),
.BI /usr/share/doc/ fte/contents.html
.SH AUTHORS
This manual page was written by Riku Voipio and Zdenek Kabelac
<riku.voipio@iki.fi> <kabi@debian.org> for the Debian GNU/Linux system.
.SH DATE
18 Nov 1999


trunk/src/fte.cpp
/*    fte.cpp
 *
 *    Copyright (c) 1994-1997, Marko Macek
 *
 *    You may distribute under the terms of either the GNU General Public
 *    License or the Artistic License, as specified in the README file.
 *
 */
#include "fte.h"
#include "log.h"
#include "c_history.h"
#ifdef USE_LOCALE
#include <locale.h>
#endif

#if defined(UNIX)
/* default locations for the configuration files */
const char *Unix_RCPaths[] = {
    "/usr/local/etc/efte/system.fterc",
    "/etc/efte/system.fterc",
    "/usr/X11R6/lib/X11/efte/system.fterc",
};

// variables used by vfte
uid_t effuid;
gid_t effgid;
#endif /* UNIX */

char ConfigFileName[MAXPATH] = "";

static void Usage() {
    printf("Usage: " PROGRAM " [-?] [-h] [--help] [-CDHTmlrt] files...\n"
           "Version: " VERSION " " COPYRIGHT "\n"
           "   You may distribute under the terms of either the GNU General Public\n"
           "   License or the Artistic License, as specified in the README file.\n"
           "\n"
           "Options:\n"
           "  --                End of options, only files remain.\n"
           "  -+                Next option is file.\n"
           "  -? -h --help      Display usage.\n"
           "  -!                Ignore config file, use builtin defaults (also -c).\n"
           "  -C[<.cnf>]        Use specified configuration file (no arg=builtin).\n"
           "  -D[<.dsk>]        Load/Save desktop from <.dsk> file (no arg=disable desktop).\n"
           "  -H[<.his>]        Load/Save history from <.his> file (no arg=disable history).\n"
           "  -m[<mode>]        Override mode for remaining files (no arg=no override).\n"
           "  -l<line>[,<col>]  Go to line (and column) in next file.\n"
           "  -r                Open next file as read-only.\n"
           "  -T[<tagfile>]     Load tags file at startup.\n"
           "  -t<tag>           Locate specified tag.\n"
//           "       -p        Load files into already running eFTE.\n"
        );
}

#ifndef UNIX
/*
 * findPathExt() returns a ^ to the suffix in a file name string. If the
 * name contains a suffix, the pointer ^ts to the suffix' dot character,
 * if the name has no suffix the pointer points to the NUL terminator of
 * the file name string.
 * .lib: CBASE.LIB
 */
static char *findPathExt(char *filename) {
    char *p, *sps;

    for (p = filename, sps = NULL; *p; p++) {
        if (ISSLASH(*p))
            sps = NULL;
        if (*p == '.')
            sps = p;
    }
    if (sps == NULL)
        sps = p;
    return sps;
}
#endif

#if defined(NT) && defined(MSVC) && !defined(__WATCOMC__)
char *getProgramName(char *name) {
    return _pgmptr;
}
#endif

#if defined(OS2) && defined(__EMX__)

// argv[0] on emx does not contain full path

#define INCL_DOS
#include <os2.h>

char *getProgramName(char *name) {
    char ProgramName[MAXPATH];
    PTIB tib;
    PPIB pib;

    DosGetInfoBlocks(&tib, &pib);
    if (DosQueryModuleName(pib->pib_hmte, sizeof(ProgramName), ProgramName) != 0)
        return name;
    return strdup(ProgramName);
}

#endif

static int GetConfigFileName(int /*argc*/, char **argv, char *ConfigFileName) {
    // NOTE: function assumes that ConfigFileName's size is MAXPATH

    char CfgName[MAXPATH] = "";

    if (ConfigFileName[0] == 0) {
#if defined(UNIX)
        // ? use ./.efterc if by current user ?
        ExpandPath("~/.efterc", CfgName, sizeof(CfgName));
#elif defined(OS2) || defined(NT)
        char home[MAXPATH] = "";
        char *ph;
#if defined(OS2)
        ph = getenv("HOME");
        if (ph) strlcpy(home, ph, sizeof(home));
#endif
#if defined(NT)
        ph = getenv("HOMEDRIVE");
        if (ph) strlcpy(home, ph, sizeof(home));
        ph = getenv("HOMEPATH");
        if (ph) strlcat(home, ph, sizeof(home));
#endif
        if (home[0]) {
            strlcpy(CfgName, home, sizeof(CfgName));
            Slash(CfgName, 1);
            strlcat(CfgName, "efte.cnf", sizeof(CfgName));
        }

        if (!home[0] || access(CfgName, 0) != 0) {
            strlcpy(CfgName, argv[0], sizeof(CfgName));

            char *extPtr;

            if ((extPtr = findPathExt(CfgName)) != NULL)
            {
                *extPtr = 0;
                strlcat(CfgName, ".cnf", sizeof(CfgName));
            }
        }
#endif

        strlcpy(ConfigFileName, CfgName, MAXPATH);
    }
    if (access(ConfigFileName, 0) == 0)
        return 1;

#if defined(UNIX)
    for (unsigned int i = 0; i < sizeof(Unix_RCPaths)/sizeof(Unix_RCPaths[0]); i++) {
        if (access(Unix_RCPaths[i], 0) == 0) {
            strlcpy(ConfigFileName, Unix_RCPaths[i], MAXPATH);
            return 1;
        }
    }
#endif
    return 0;
}

static int CmdLoadConfiguration(int &argc, char **argv) {
    int ign = 0;
    int QuoteAll = 0, QuoteNext = 0;
    int haveConfig = 0;
    int Arg;

    for (Arg = 1; Arg < argc; Arg++) {
        if (!QuoteAll && !QuoteNext && (argv[Arg][0] == '-')) {
            if (argv[Arg][1] == '-') {
                if (strcmp(argv[Arg], "--help") == 0) {
                    Usage();
                    return 0;
                }
                int debug_clean = strcmp(argv[Arg], "--debugclean") == 0;
                if (debug_clean || strcmp(argv[Arg], "--debug") == 0) {
#ifndef FTE_NO_LOGGING
                    char path[MAXPATH];
#ifdef UNIX
                    ExpandPath("~/.efte", path, sizeof(path));
#else
                    JustDirectory(argv[0], path, sizeof(path));
#endif
                    Slash(path,1);
                    strlcat(path, "efte.log", sizeof(path));
                    if (debug_clean) unlink(path);

                    globalLog.SetLogFile(path);
                    printf("Trace Log in: %s\n", path);
#else
                    printf("--debug, --debugclean disabled\n");
#endif
                }
                else
                    QuoteAll = 1;
            } else if (argv[Arg][1] == '!') {
                ign = 1;
            } else if (argv[Arg][1] == '+') {
                QuoteNext = 1;
            } else if (argv[Arg][1] == '?' || argv[Arg][1] == 'h') {
                Usage();
                return 0;
            } else if (argv[Arg][1] == 'c' || argv[Arg][1] == 'C') {
                if (argv[Arg][2])
                {
                    ExpandPath(argv[Arg] + 2, ConfigFileName, sizeof(ConfigFileName));
                    haveConfig = 1;
                }
                else
                    ign = 1;
            }
        }
    }
    if (!haveConfig && GetConfigFileName(argc, argv, ConfigFileName) == 0) {
        // should we default to internal
#ifdef DEFAULT_INTERNAL_CONFIG
       ign = 1;
#endif
    }

    if (ign) {
        if (UseDefaultConfig() == -1)
            DieError(1, "Error in internal configuration??? FATAL!");
    } else {
        if (LoadConfig(argc, argv, ConfigFileName) == -1)
            DieError(1,
                     "Failed to load configuration file '%s'.\n"
                     "Use '-C' option.", ConfigFileName);
    }
    for (Arg = 1; Arg < argc; Arg++) {
        if (!QuoteAll && !QuoteNext && (argv[Arg][0] == '-')) {
            if (argv[Arg][1] == '-' && argv[Arg][2] == '\0') {
                QuoteAll = 1;
            } else if (argv[Arg][1] == '+') {
                QuoteNext = 1;
            } else if (argv[Arg][1] == 'D') {
                ExpandPath(argv[Arg] + 2, DesktopFileName, sizeof(DesktopFileName));
                if (IsDirectory(DesktopFileName)) {
                    Slash(DesktopFileName, 1);
                    strlcat(DesktopFileName, DESKTOP_NAME, sizeof(DesktopFileName));
                }
                if (DesktopFileName[0] == 0) {
                    LoadDesktopOnEntry = 0;
                    SaveDesktopOnExit = 0;
                } else {
                    LoadDesktopOnEntry = 1;
                }
            } else if (argv[Arg][1] == 'H') {
                strlcpy(HistoryFileName, argv[Arg] + 2, sizeof(HistoryFileName));
                if (HistoryFileName[0] == 0) {
                    KeepHistory = 0;
                } else {
                    KeepHistory = 1;
                }
            }
        } else {
            if (LoadDesktopOnEntry == 2) {
                LoadDesktopOnEntry = 0;
                SaveDesktopOnExit = 0;
                DesktopFileName[0] = 0;
            }
        }
    }
    if (LoadDesktopOnEntry == 2)
        LoadDesktopOnEntry = 1;
    return 1;
}

int main(int argc, char **argv) {
#if defined(_DEBUG) && defined(MSVC) && defined(MSVCDEBUG)
   _CrtSetReportMode( _CRT_WARN, _CRTDBG_MODE_FILE );
   _CrtSetReportFile( _CRT_WARN, _CRTDBG_FILE_STDERR );
   _CrtSetReportMode( _CRT_ERROR, _CRTDBG_MODE_FILE );
   _CrtSetReportFile( _CRT_ERROR, _CRTDBG_FILE_STDERR );
   _CrtSetReportMode( _CRT_ASSERT, _CRTDBG_MODE_FILE );
   _CrtSetReportFile( _CRT_ASSERT, _CRTDBG_FILE_STDERR );
#endif //_DEBUG && MSVC && MSVCDEBUG

#if defined(__EMX__) || (defined(NT) && defined(MSVC) && !defined(__WATCOMC__))
    argv[0] = getProgramName(argv[0]);
#endif

#if defined(UNIX) && defined(LINUX)
    // security fix - when we need to be suid to access vcsa
    effuid = geteuid();
    effgid = getegid();

    if (getuid() != effuid)
        seteuid(getuid());
    if (getgid() != effgid)
        setegid(getgid());
#endif

#ifdef USE_LOCALE
    // setup locale from environment
    setlocale(LC_ALL, "");
#endif

    if (CmdLoadConfiguration(argc, argv) == 0)
        return 1;

    STARTFUNC("main");

    EGUI *g = new EGUI(argc, argv, ScreenSizeX, ScreenSizeY);
    if (gui == 0 || g == 0)
        DieError(1, "Failed to initialize display\n");

    gui->Run();

#if defined(OS2) && !defined(DBMALLOC) && defined(CHECKHEAP)
    if (_heapchk() != _HEAPOK)
        DieError(0, "Heap memory is corrupt.");
#endif

    delete gui;
    gui = 0;

#if defined(__EMX__)
    free(argv[0]);
#endif

#if defined(OS2) && !defined(DBMALLOC) && defined(CHECKHEAP)
    if (_heapchk() != _HEAPOK)
        DieError(0, "Heap memory is corrupt.");
#endif

#if defined(_DEBUG) && defined(MSVC) && defined(MSVCDEBUG)
    _CrtSetDbgFlag((_CRTDBG_LEAK_CHECK_DF) | _CrtSetDbgFlag(_CRTDBG_REPORT_FLAG));
#endif //_DEBUG && MSVC && MSVCDEBUG

#if defined(__DEBUG_ALLOC__)
    _dump_allocated(64);
#endif

    ENDFUNCRC(0);
    //return 0;
}
